File: transform/Swift/Sources/Transform/Transformclient.swift
// Copyright Core DF — Apache License 2.0

import Foundation

public enum Transformclient {
    public static func JsonParse(_ text: String) -> [String: Any] {
        guard let data = text.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) else {
            return ["status_code": 400, "error": "json parse error"]
        }
        return ["status_code": 200, "data": obj]
    }

    public static func JsonStringify(_ data: Any, indent: Int? = nil) -> [String: Any] {
        guard JSONSerialization.isValidJSONObject(data) || data is String || data is NSNull else {
            return ["status_code": 400, "error": "invalid data"]
        }
        do {
            let opts: JSONSerialization.WritingOptions = indent != nil ? [.prettyPrinted] : []
            let d = try JSONSerialization.data(withJSONObject: data, options: opts)
            let text = String(data: d, encoding: .utf8) ?? ""
            return ["status_code": 200, "text": text]
        } catch {
            return ["status_code": 400, "error": error.localizedDescription]
        }
    }

    public static func CsvToRows(_ text: String, delimiter: String = ",") -> [String: Any] {
        let lines = text.split(whereSeparator: \.isNewline).map(String.init).filter { !$0.isEmpty }
        guard let header = lines.first else { return ["status_code": 400, "error": "empty csv"] }
        let cols = header.split(separator: Character(delimiter.first ?? ",")).map(String.init)
        var rows: [[String: String]] = []
        for line in lines.dropFirst() {
            let vals = line.split(separator: Character(delimiter.first ?? ",")).map(String.init)
            var row: [String: String] = [:]
            for (i, c) in cols.enumerated() { row[c] = i < vals.count ? vals[i] : "" }
            rows.append(row)
        }
        return ["status_code": 200, "rows": rows]
    }

    public static func RowsToCsv(_ rows: [[String: String]], delimiter: String = ",") -> [String: Any] {
        guard let first = rows.first, !first.isEmpty else {
            return ["status_code": 400, "error": "rows must not be empty"]
        }
        let keys = Array(first.keys)
        var out = keys.joined(separator: delimiter) + "\n"
        for r in rows {
            out += keys.map { r[$0] ?? "" }.joined(separator: delimiter) + "\n"
        }
        return ["status_code": 200, "text": out]
    }

    public static func XmlToDict(_ text: String) -> [String: Any] {
        guard let tag = text.range(of: "<(\\w+)", options: .regularExpression) else {
            return ["status_code": 400, "error": "xml parse error"]
        }
        let name = String(text[text.index(tag.lowerBound, offsetBy: 1)...].prefix { $0 != ">" && $0 != " " })
        return ["status_code": 200, "data": [name: [:] as Any]]
    }

    public static func DictToXml(_ data: [String: Any], rootTag: String = "root") -> [String: Any] {
        ["status_code": 200, "text": "<\(rootTag)/>"]
    }
}

File: transform/Kotlin/src/main/kotlin/com/coredf/transform/TransformClient.kt
// Copyright Core DF
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.coredf.transform
import org.w3c.dom.Element
import java.io.ByteArrayInputStream
import java.io.StringWriter
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.transform.TransformerFactory
import javax.xml.transform.dom.DOMSource
import javax.xml.transform.stream.StreamResult

object TransformClient {
    @JvmStatic fun JsonParse(text: String) = try { Result.ok(mapOf("data" to JsonUtil.parse(text))) } catch (e: Exception) { Result.error(400, e.message) }
    @JvmStatic fun JsonStringify(data: Any?, indent: Int? = null) = try { Result.ok(mapOf("text" to JsonUtil.stringify(data))) } catch (e: Exception) { Result.error(400, e.message) }
    @JvmStatic fun CsvToRows(text: String, delimiter: String = ",") = try {
        val lines = text.split("\r?\n".toRegex())
        if (lines.isEmpty()) return Result.ok(mapOf("rows" to emptyList<Map<String,String>>()))
        val headers = lines[0].split(delimiter)
        val rows = lines.drop(1).filter { it.isNotEmpty() }.map { line ->
            val cols = line.split(delimiter)
            headers.mapIndexed { i, h -> h to (cols.getOrNull(i) ?: "") }.toMap()
        }
        Result.ok(mapOf("rows" to rows))
    } catch (e: Exception) { Result.error(400, e.message) }

    @JvmStatic fun RowsToCsv(rows: List<Map<String, String>>, delimiter: String = ","): Result {
        if (rows.isEmpty()) return Result.error(400, "rows must not be empty")
        val headers = rows[0].keys.toList()
        val sb = StringBuilder(headers.joinToString(delimiter)).append('\n')
        rows.forEach { row -> sb.append(headers.joinToString(delimiter) { row[it] ?: "" }).append('\n') }
        return Result.ok(mapOf("text" to sb.toString()))
    }

    @JvmStatic fun XmlToDict(text: String) = try {
        val doc = DocumentBuilderFactory.newInstance().newDocumentBuilder().parse(ByteArrayInputStream(text.toByteArray()))
        val root = doc.documentElement; Result.ok(mapOf("data" to mapOf(root.tagName to elem(root))))
    } catch (e: Exception) { Result.error(400, e.message) }

    @JvmStatic fun DictToXml(data: Map<String, Any?>, rootTag: String = "root") = try {
        val doc = DocumentBuilderFactory.newInstance().newDocumentBuilder().newDocument()
        val root = doc.createElement(rootTag); doc.appendChild(root)
        data.forEach { (k, v) -> build(root, v, k) }
        val sw = StringWriter(); TransformerFactory.newInstance().newTransformer().transform(DOMSource(doc), StreamResult(sw))
        Result.ok(mapOf("text" to sw.toString()))
    } catch (e: Exception) { Result.error(400, e.message) }

    private fun elem(node: Element): Any {
        val children = (0 until node.childNodes.length).mapNotNull { i -> node.childNodes.item(i) as? Element }
        if (children.isEmpty()) return node.textContent?.trim() ?: ""
        val out = linkedMapOf<String, Any>()
        children.forEach { child ->
            val v = elem(child); val tag = child.tagName
            if (out.containsKey(tag)) {
                val prev = out[tag]; out[tag] = if (prev is MutableList<*>) { (prev as MutableList<Any>).apply { add(v) }; prev } else mutableListOf(prev!!, v)
            } else out[tag] = v
        }
        return out
    }
    @Suppress("UNCHECKED_CAST")
    private fun build(parent: org.w3c.dom.Element, obj: Any?, tag: String) {
        val doc = parent.ownerDocument
        when (obj) {
            is Map<*, *> -> { val node = doc.createElement(tag); parent.appendChild(node); obj.forEach { (k, v) -> build(node, v, k.toString()) } }
            is List<*> -> obj.forEach { build(parent, it, tag) }
            else -> { val node = doc.createElement(tag); node.textContent = obj?.toString() ?: ""; parent.appendChild(node) }
        }
    }
}
